File: app/src/main/java/com/ndmsystems/coala/observer/ObservingResource.java
package com.ndmsystems.coala.observer;

import com.ndmsystems.coala.CoAPHandler;
import com.ndmsystems.coala.message.CoAPMessage;
import com.ndmsystems.infrastructure.logging.LogHelper;

/**
 * Created by bas on 14.11.16.
 */

public class ObservingResource {
    private Long validUntil = System.currentTimeMillis() + 30000;
    private Integer sequenceNumber = -1;

    private final CoAPHandler handler;
    private final CoAPMessage initiatingMessage;

    public ObservingResource(CoAPMessage initiatingMessage, CoAPHandler handler) {
        this.handler = handler;
        this.initiatingMessage = initiatingMessage;
    }

    public Integer getSequenceNumber() {
        return sequenceNumber;
    }

    public void setSequenceNumber(Integer sequenceNumber) {
        this.sequenceNumber = sequenceNumber;
    }

    public String getUri() {
        return initiatingMessage.getURI();
    }

    public CoAPMessage getInitiatingMessage() {
        return initiatingMessage;
    }

    public CoAPHandler getHandler() {
        return handler;
    }

    public void setMaxAge(Integer maxAge) {
        LogHelper.v("Set max age at " + maxAge);
        validUntil = System.currentTimeMillis() + maxAge * 1000;
    }

    public boolean isExpired() {
        LogHelper.d("is resource (" + initiatingMessage.getURIPathString() + ") expired? " + (System.currentTimeMillis() >= validUntil));
        return System.currentTimeMillis() >= validUntil;
    }
}

File: app/src/main/java/com/ndmsystems/coala/observer/ObservingResource.kt
package com.ndmsystems.coala.observer

import com.ndmsystems.coala.CoAPHandler
import com.ndmsystems.coala.helpers.MonotonicClock
import com.ndmsystems.coala.helpers.logging.LogHelper.d
import com.ndmsystems.coala.helpers.logging.LogHelper.v
import com.ndmsystems.coala.message.CoAPMessage

/**
 * @param clock seam for tests: the renewal deadline is measured against this rather than the system
 * clock, so a test can age a subscription instead of asking for a max age of zero and hoping.
 */
class ObservingResource(
    val initiatingMessage: CoAPMessage,
    val handler: CoAPHandler?,
    private val clock: MonotonicClock = MonotonicClock.SYSTEM
) {
    private var validUntil = clock.nowMillis() + DEFAULT_MAX_AGE_MILLIS
    var sequenceNumber = -1

    val uri: String
        get() = initiatingMessage.getURI()

    fun setMaxAge(maxAge: Int) {
        v("Set max age at $maxAge")
        // Max-Age is a peer-controlled uint32, and the option decoder hands it over as a signed
        // Int - a value above Int.MAX_VALUE arrives negative and would put the deadline in the
        // past, expiring the subscription the moment the peer asked for the longest one. Read the
        // bits back as unsigned, and multiply in Long so the seconds->millis scaling cannot wrap.
        validUntil = clock.nowMillis() + (maxAge.toLong() and 0xFFFFFFFFL) * 1000L
    }

    val isExpired: Boolean
        get() {
            d("is resource (" + initiatingMessage.getURIPathString() + ") expired? " + (clock.nowMillis() >= validUntil))
            return clock.nowMillis() >= validUntil
        }

    private companion object {
        /** What a subscription is assumed good for until the peer says otherwise. */
        const val DEFAULT_MAX_AGE_MILLIS = 30_000L
    }
}